---
title: "ABP1_swap_gained_interaction"
author: "David Bradley"
date: "1/21/2020"
output: html_document
---

```{r}

# Script to analyse PCA  data

library(tidyverse)
library(readxl)
library(seqinr)

# Interactions gained in ABP1 after domain swap
sh3_pca <- read_csv('PCAOct18ABP1SWAPIntGained_0120Analysis.csv', skip=0, col_names=TRUE)
interactors_ids <- unique(unlist(sh3_pca[,2]))

# Write out the IDs
write.table(interactors_ids,file='abp1_swap_gained_pca.txt',col.names=FALSE, row.names=FALSE, quote=FALSE)

# Use UniProt to manually retrieve the sequences from the Ensembl IDs (https://www.uniprot.org/uploadlists/)

# Read in the interactor sequences
interactors_fasta <- read.fasta('AB1_swap_gained_interactors.fasta',seqtype='AA')
interactors_seq <- getSequence(interactors_fasta)

accession <- rapply(strsplit(names(interactors_fasta),split='\\|'), function (x) x[2])
gene_name <- rapply(strsplit(rapply(strsplit(unlist(getAnnot(interactors_fasta)),split='GN='), function (x) x[2]),split=' '),function(x) x[1])
interactors_name <- paste(interactors_ids,accession,gene_name,sep='_')
names(interactors_seq) <- interactors_name

# Read through each PWM file to generate a list of PWMs

pwm_list <- NULL

for (i in 1:length(pwm_files)) {
  
  path <- paste('/DoMo-Pred/pwm_dir/',pwm_files[i],sep='')
  
  pwm <- read.table(path)
  rownames(pwm) <- pwm[,1]
  colnames(pwm) <- NULL
  pwm <- pwm[,-1]
  
  pwm_list <- c(pwm_list, list(pwm))
  
}

names(pwm_list) <- pwm_files

# Now, the PWM IDs need to be mapped to gene names
# # These PWMs come from (PMID: 26861823), who sourced them from (PMID: 19841731).

pwm_ids <- rapply(strsplit(rapply(strsplit(names(pwm_list),split='_'), function(x) x[1]),split='-'), function(x) x[1])

# Again, out the ID file and then retrieve the gene names manually using the Uniprot ID mapper
# (https://www.uniprot.org/uploadlists/)

write.table(pwm_ids,file='sh3_pwm_accessions_uniprot.txt',col.names=FALSE, row.names=FALSE, quote=FALSE)

# Read in mapping table and process the strings to generate new names for the PWMs

yeast_sh3_map <- read.table('yeast_sh3_accession_to_GN.txt',header=TRUE,stringsAsFactors = FALSE)
id_match <- yeast_sh3_map[match(pwm_ids,yeast_sh3_map[,1]),2]
pwm_names <- paste(id_match,substr(names(pwm_list),7,rapply(strsplit(names(pwm_list),split=''), function(x) length(x))),sep='')

sh3_yeast <- unname(rapply(strsplit(unlist(sh3_pca[,1]),split=' '), function(x) x[1]))
names(pwm_list) <- rapply(strsplit(pwm_names,split='_'), function(x) x[1])

```

MSS function to score the similarity between SH3 domains and candidate SH3 ligands

```{r}

mss_score <- function(psite,pwm) {
  
  psite <- unlist(strsplit(psite,split=''))
  pwm <- as.matrix(pwm)
    
  match.ic = apply(pwm, 2, function(col) sum(col * logb(20 * col), na.rm=T))
  
  
  c=0
  min=0
  max = 0
  
  for (i in 1:ncol(pwm)) {
    
    aa <- psite[i]  
    
    # For missing amino acids, score the '_' position
    # to be the same as a minimum frequency AA
    
    if (aa == '_') {
      current <- min(pwm[,i])*match.ic[i]
    } else {
      current <- pwm[rownames(pwm) %in% psite[i],i]*match.ic[i]
    }
    
    minimum <- min(pwm[,i])*match.ic[i]
    maximum <- max(pwm[,i])*match.ic[i]
    
    c = c+current
    min = min+minimum
    max = max+maximum
    
  }
  
  mss = (c-min)/(max-min)
  
  return(unname(mss))
  
}

```

Now, the idea is to iterate through each one of the gained interactions, and then find the
maximum MSS for the corresponding SH3 PWM

```{r}

max_mss_list <- NULL
tag_vec <- NULL

for (i in 1:length(pwm_list)) {
  
  print(i)
  
  pwm <- pwm_list[[i]]
  id <- names(pwm_list)[i]
  
  pwm_targets <- NULL
  target_seqs <- NULL
  
  # Locate the interactors that are gained when a new SH3 is swapped into Abp1
  pwm_targets <- unique(unname(unlist(sh3_pca[sh3_yeast %in% id,2])))
  
  if(length(pwm_targets) == 0) {next}
  
  # Retrieve the sequences corresponding to the gained interactions
  target_seqs <- interactors_seq[rapply(strsplit(names(interactors_seq),split='_'), function(x) x[1]) %in% pwm_targets]
  
  max_mss_vec <- NULL
  
  # Calculate MSS scores for all possible candidate sites
  for (j in 1:length(pwm_targets)) {
    
    tag <- paste(names(pwm_list)[i],'_',names(target_seqs)[j],sep='')
    
    target_seq <- target_seqs[[j]]
    
    start_site_vec <- NULL
    end_site_vec <- NULL
    
    # Generate the start and end sites
    
    for (k in 1:ncol(pwm)) {
      
      start_site <- paste(paste(rep('_',ncol(pwm)-k),collapse=''),paste(target_seq[1:k],collapse=''),sep='')
      start_site_vec <- c(start_site_vec, start_site)
      end_site <- paste(paste(target_seq[(length(target_seq)-k+1):length(target_seq)],collapse=''),paste(rep('_',ncol(pwm)-k),collapse=''),sep='')
      end_site_vec <- c(end_site_vec, end_site)  
      
    }
    
    # Now extract all possile binding sites from the sequence
    
    middle_site_vec <- NULL
    
    for (k in 2:(length(target_seq)-k)) {
      
      middle_site <- paste(target_seq[k:(k+ncol(pwm)-1)],collapse='')
      middle_site_vec <- c(middle_site_vec, middle_site)
      
    }
    
    candidate_sites <- NULL
    candidate_sites <- c(start_site_vec,middle_site_vec,rev(end_site_vec))
    
    # Iterate through all of the candidate sites and score with the MSS
    
    mss_vec <- NULL
    
    for (k in 1:length(candidate_sites)) {
      
      mss <- mss_score(candidate_sites[k],pwm)
      mss_vec <- c(mss_vec,mss)
      
    }
    
    max_mss <- round(as.numeric(max(mss_vec)),3)
    max_mss_vec <- c(max_mss_vec, max_mss)
    tag_vec <- c(tag_vec,tag)
  }
  
  max_mss_mat <- cbind(max_mss_vec,pwm_targets)
  
  max_mss_list <- c(max_mss_list, list(max_mss_mat))
  names(max_mss_list)[length(max_mss_list)] <- id
}

real_data <- as.numeric(unname(rapply(max_mss_list,function(x) x[,1])))

# Correct for the fact that some SH3 domains are represented by more than one PWM (e.g. class I and class II)

duplicated <- tag_vec[which(duplicated(tag_vec) == TRUE)]
dup_score_max_vec <- NULL

for (i in 1:length(duplicated)) {
  
  id <- duplicated[i] 
  
  dup_scores <- real_data[tag_vec %in% id]
  
  dup_score_max_vec <- c(dup_score_max_vec,max(dup_scores))
  
}

real_data <- real_data[!(tag_vec %in% duplicated)]
real_data <- c(real_data, dup_score_max_vec)

```

Perform the same procedure as above but for random sites

```{r}

# Calculate bg AA frequencies

library(seqinr)

AAs <- c('A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y')

sc_proteome <- read.fasta('Scerevisiae_uniprot_reference.fasta', seqtype='AA')
sc_aa_tab <- table(unlist(getSequence(sc_proteome)))
sc_aa_tab <- sc_aa_tab[match(AAs,names(sc_aa_tab))]
sc_aa_freq <- sc_aa_tab/sum(sc_aa_tab)

random_data_vec <- NULL

for (z in 1:50) {
  
  pseudo_max_mss_list_x100 <- NULL
  
  tag_vec <- NULL
  
  print(z)
  
  for (i in 1:length(pwm_list)) {
  
  #print(i)
  
  pwm <- pwm_list[[i]]
  id <- names(pwm_list)[i]
  
  pwm_targets <- NULL
  target_seqs <- NULL
  
  # Retrieve the SH3 domain interactors
  pwm_targets <- unique(unname(unlist(sh3_pca[sh3_yeast %in% id,2])))
  
  if(length(pwm_targets) == 0) {next}
  
  # Retrieve the sequences of the SH3 domain interactors
  target_seqs <- interactors_seq[rapply(strsplit(names(interactors_seq),split='_'), function(x) x[1]) %in% pwm_targets]
  
  max_mss_vec <- NULL
  
  #Calculate MSS scores for all candidate sites
  for (j in 1:length(pwm_targets)) {
    
    target_seq <- target_seqs[[j]]
    tag <- paste(names(pwm_list)[i],'_',names(target_seqs)[j],sep='')
    
    # Random sequence
    
    target_seq <- sample(AAs, prob = sc_aa_freq, size = length(target_seq), replace=TRUE)
    start_site_vec <- NULL
    end_site_vec <- NULL
    
    # Generate the start and end sites
    
    for (k in 1:ncol(pwm)) {
      
      start_site <- paste(paste(rep('_',ncol(pwm)-k),collapse=''),paste(target_seq[1:k],collapse=''),sep='')
      start_site_vec <- c(start_site_vec, start_site)
      end_site <- paste(paste(target_seq[(length(target_seq)-k+1):length(target_seq)],collapse=''),paste(rep('_',ncol(pwm)-k),collapse=''),sep='')
      end_site_vec <- c(end_site_vec, end_site)  
      
    }
    
    # Now extract all possile binding sites from the sequence
    
    middle_site_vec <- NULL
    
    for (k in 2:(length(target_seq)-k)) {
      
      middle_site <- paste(target_seq[k:(k+ncol(pwm)-1)],collapse='')
      middle_site_vec <- c(middle_site_vec, middle_site)
      
    }
    
    candidate_sites <- NULL
    candidate_sites <- c(start_site_vec,middle_site_vec,rev(end_site_vec))
    
    # Iterate through all of the candidate sites and score with the MSS
    
    mss_vec <- NULL
    
    for (k in 1:length(candidate_sites)) {
      
      mss <- mss_score(candidate_sites[k],pwm)
      mss_vec <- c(mss_vec,mss)
      
    }
    
    max_mss <- round(as.numeric(max(mss_vec)),3)
    max_mss_vec <- c(max_mss_vec, max_mss)
    tag_vec <- c(tag_vec,tag)
  }
  
  max_mss_mat <- cbind(max_mss_vec,pwm_targets)
  
  pseudo_max_mss_list_x100 <- c(pseudo_max_mss_list_x100, list(max_mss_mat))
  names(pseudo_max_mss_list_x100)[length(pseudo_max_mss_list_x100)] <- id
  
}

  random_data <- as.numeric(unname(rapply(pseudo_max_mss_list_x100,function(x) x[,1])))

  # Correct for the fact that some SH3 domains are represented by more than one PWM (e.g. Class I and   # Class II)

  duplicated <- tag_vec[which(duplicated(tag_vec) == TRUE)]

  dup_score_max_vec <- NULL

  for (d in 1:length(duplicated)) {
  
    id <- duplicated[d] 
    dup_scores <- random_data[tag_vec %in% id]
    dup_score_max_vec <- c(dup_score_max_vec,max(dup_scores))
  
  }

  random_data <- random_data[!(tag_vec %in% duplicated)]
  random_data <- c(random_data, dup_score_max_vec)
  random_data_vec <- c(random_data_vec,random_data)
  
  
}  

# Sample N=10,000

pseudo_data <- sample(random_data_vec,size=10000,replace=FALSE)

```

Repeat the procedures above but this time scoring sites in ABP1 that do not change
upon domain swap with another SH3 domain

```{r}

library(readxl)
library(seqinr)

# Interactions in ABP1 unaffected by the domain swap
sh3_pca <- read_csv('PCAOct18ABP1SWAPIntUnAffected_0120Analysis.csv', skip=0, col_names=TRUE)

interactors_ids <- unique(unlist(sh3_pca[,2]))

write.table(interactors_ids,file='abp1_swap_retained_pca.txt',col.names=FALSE, row.names=FALSE, quote=FALSE)

# Use UniProt to retrieve the sequences from the Ensembl IDs (https://www.uniprot.org/uploadlists/)

interactors_fasta <- read.fasta('AB1_swap_retained_interactors.fasta',seqtype='AA')
interactors_seq <- getSequence(interactors_fasta)

accession <- rapply(strsplit(names(interactors_fasta),split='\\|'), function (x) x[2])
gene_name <- rapply(strsplit(rapply(strsplit(unlist(getAnnot(interactors_fasta)),split='GN='), function (x) x[2]),split=' '),function(x) x[1])
interactors_name <- paste(interactors_ids,accession,gene_name,sep='_')
names(interactors_seq) <- interactors_name

sh3_yeast <- unname(rapply(strsplit(unlist(sh3_pca[,1]),split=' '), function(x) x[1]))

```  

Iterate through each of the unaffected interactions and calculate the maximum MSS

```{r}

max_mss_list <- NULL
tag_vec <- NULL

for (i in 1:length(pwm_list)) {
  
  print(i)
  
  pwm <- pwm_list[[i]]
  id <- names(pwm_list)[i]
  
  pwm_targets <- NULL
  target_seqs <- NULL
  
  # Retrieve the IDs of the SH3 domain interactors
  pwm_targets <- unique(unname(unlist(sh3_pca[sh3_yeast %in% id,2])))
  
  if(length(pwm_targets) == 0) {next}
  
  # Retrieve the corresponding sequences of these interactors
  target_seqs <- interactors_seq[rapply(strsplit(names(interactors_seq),split='_'), function(x) x[1]) %in% pwm_targets]
  
  max_mss_vec <- NULL
  
  # Calculate the MSS for all possible candidate sites
  for (j in 1:length(pwm_targets)) {
    
    #
    tag <- paste(names(pwm_list)[i],'_',names(target_seqs)[j],sep='')
    target_seq <- target_seqs[[j]]
    start_site_vec <- NULL
    end_site_vec <- NULL
    
    # Generate the start and end sites
    
    for (k in 1:ncol(pwm)) {
      
      start_site <- paste(paste(rep('_',ncol(pwm)-k),collapse=''),paste(target_seq[1:k],collapse=''),sep='')
      start_site_vec <- c(start_site_vec, start_site)
      end_site <- paste(paste(target_seq[(length(target_seq)-k+1):length(target_seq)],collapse=''),paste(rep('_',ncol(pwm)-k),collapse=''),sep='')
      end_site_vec <- c(end_site_vec, end_site)  
      
    }
    
    # Now extract all possile binding sites from the sequence
    
    middle_site_vec <- NULL
    
    for (k in 2:(length(target_seq)-k)) {
      
      middle_site <- paste(target_seq[k:(k+ncol(pwm)-1)],collapse='')
      middle_site_vec <- c(middle_site_vec, middle_site)
      
    }
    
    candidate_sites <- NULL
    candidate_sites <- c(start_site_vec,middle_site_vec,rev(end_site_vec))
    
    # Iterate through all of the candidate sites and score with the MSS
    
    mss_vec <- NULL
    
    for (k in 1:length(candidate_sites)) {
      
      mss <- mss_score(candidate_sites[k],pwm)
      mss_vec <- c(mss_vec,mss)
      
    }
    
    max_mss <- round(as.numeric(max(mss_vec)),3)
    max_mss_vec <- c(max_mss_vec, max_mss)
    tag_vec <- c(tag_vec,tag)
  }
  
  max_mss_mat <- cbind(max_mss_vec,pwm_targets)
  
  max_mss_list <- c(max_mss_list, list(max_mss_mat))
  names(max_mss_list)[length(max_mss_list)] <- id
}

retained_data <- as.numeric(unname(rapply(max_mss_list,function(x) x[,1])))

# Correct for the fact that some SH3 domains have more than one PWM (e.g. ClassI and ClassII)

duplicated <- tag_vec[which(duplicated(tag_vec) == TRUE)]

dup_score_max_vec <- NULL

for (i in 1:length(duplicated)) {
  
  id <- duplicated[i] 
  dup_scores <- retained_data[tag_vec %in% id]
  dup_score_max_vec <- c(dup_score_max_vec,max(dup_scores))
  
}

retained_data <- retained_data[!(tag_vec %in% duplicated)]
retained_data <- c(retained_data, dup_score_max_vec)

```

GG boxplot of the data

```{r}

mss_data <- cbind(c(rep('Random',10000),rep('Int. retained',1281),rep('Int. gained',317)),c(pseudo_data,retained_data,real_data))
mss_data <- data.frame(mss_data,stringsAsFactors = FALSE)
mss_data[,2] <- as.numeric(mss_data[,2])
colnames(mss_data) <- c('Protein','Max_MSS')

mss_data$Protein <- factor(mss_data$Protein, levels = c('Random','Int. retained','Int. gained'))

p <- ggplot(mss_data, aes(x=Protein, y=Max_MSS, color=Protein)) + geom_boxplot(color=c(colors()[261],colors()[261],colors()[261]),fill=c('White','#DCDCDC','#B2182B'),lwd=0.95)
p <- p+theme_bw()
p <- p + ylab("Max. MSS") + xlab("") + ggtitle('')
p <- p+theme(axis.text=element_text(size=13),axis.title=element_text(size=13.5,face="bold"),plot.title=element_text(size=16,face='bold'))
p <- p+theme(plot.title = element_text(hjust = 0.5))
p <- p + theme(legend.position = "none") 
p <- p + scale_x_discrete(breaks=c('Random','Int. retained','Int. gained'),labels=c('Random','PPI retained','PPI gained'))
p <- p+coord_flip()
p <- p + geom_text(data=data.frame(), aes(x=c(1.3,2.3,3.3), y=c(0.9,0.9,0.9), label=c('n=10,000','n=1,281', 'n=317')), col='black', size=4.25)

ggsave(file='FigS3E_horizontal.pdf', plot=p, width=5.8, height=3.3)
ggsave(file='FigS3E_horizontal.pdf', plot=p, width=7.0, height=3.3)

```

